OIDC CodeAuthenticationMechanism does not work if the proactive authentication is disabled

**Describe the bug**

Adding `quarkus.http.auth.proactive=false` breaks the OIDC adapter code flow (ex, adding this property to `integration-tests/oidc-code-flow/..../application.properties` breaks the tests)
Reported originally at https://quarkusio.zulipchat.com/#narrow/stream/187030-users/topic/OIDC.20BearerAuthenticationMechanism.20for.20unproteced.20resources

I think it can be a simple fix, hopefully, which needs to be added somewhere around the code which deals with the disabled proactive authentication:
```
2020-09-14 18:15:22,066 ERROR [io.qua.ver.htt.run.QuarkusErrorHandler] (executor-thread-1) HTTP Request to /web-app/callback-after-redirect?state=13d4ff40-f6b1-4f83-abad-e696113135dd&session_state=d4aa68b2-6a89-4073-8060-b2eedd3887c7&code=c3ff900f-f716-46d5-a7b3-e2d155ba7b1f.d4aa68b2-6a89-4073-8060-b2eedd3887c7.3d383035-f469-4344-9402-5fbda2b1f033 failed, error id: cf969c9e-cae8-4e92-ab17-944c0bde31a1-1: org.jboss.resteasy.spi.UnhandledException: io.quarkus.vertx.http.runtime.security.AuthenticationRedirectException

	at org.jboss.resteasy.core.ExceptionHandler.handleApplicationException(ExceptionHandler.java:106)
	at org.jboss.resteasy.core.ExceptionHandler.handleException(ExceptionHandler.java:372)
	at org.jboss.resteasy.core.SynchronousDispatcher.writeException(SynchronousDispatcher.java:216)
	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:515)
	at org.jboss.resteasy.core.SynchronousDispatcher.lambda$invoke$4(SynchronousDispatcher.java:259)
	at org.jboss.resteasy.core.SynchronousDispatcher.lambda$preprocess$0(SynchronousDispatcher.java:160)
	at org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:364)
	at org.jboss.resteasy.core.SynchronousDispatcher.preprocess(SynchronousDispatcher.java:163)
	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:245)
	at io.quarkus.resteasy.runtime.standalone.RequestDispatcher.service(RequestDispatcher.java:73)
	at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.dispatch(VertxRequestHandler.java:131)
	at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.access$000(VertxRequestHandler.java:37)
	at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler$1.run(VertxRequestHandler.java:94)
	at org.jboss.threads.ContextClassLoaderSavingRunnable.run(ContextClassLoaderSavingRunnable.java:35)
	at org.jboss.threads.EnhancedQueueExecutor.safeRun(EnhancedQueueExecutor.java:2046)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.doRunTask(EnhancedQueueExecutor.java:1578)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1452)
	at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
	at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
	at java.lang.Thread.run(Thread.java:748)
	at org.jboss.threads.JBossThread.run(JBossThread.java:479)
Caused by: io.quarkus.vertx.http.runtime.security.AuthenticationRedirectException
	at io.quarkus.oidc.runtime.CodeAuthenticationMechanism$4$1.accept(CodeAuthenticationMechanism.java:326)
...
```

It is handled in the proactive auth case here:
https://github.com/quarkusio/quarkus/blob/master/extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/security/HttpSecurityRecorder.java#L84

So we probably need to add the same somewhere here:
https://github.com/quarkusio/quarkus/blob/master/extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/security/HttpSecurityRecorder.java#L156

--- a/extensions/resteasy/deployment/src/main/java/io/quarkus/resteasy/deployment/ResteasyBuiltinsProcessor.java
+++ b/extensions/resteasy/deployment/src/main/java/io/quarkus/resteasy/deployment/ResteasyBuiltinsProcessor.java
@@ -21,7 +21,10 @@ import io.quarkus.deployment.builditem.ApplicationArchivesBuildItem;
 import io.quarkus.deployment.builditem.CapabilityBuildItem;
 import io.quarkus.deployment.builditem.CombinedIndexBuildItem;
 import io.quarkus.resteasy.common.spi.ResteasyJaxrsProviderBuildItem;
+import io.quarkus.resteasy.runtime.AuthenticationCompletionExceptionMapper;
 import io.quarkus.resteasy.runtime.AuthenticationFailedExceptionMapper;
+import io.quarkus.resteasy.runtime.AuthenticationRedirectExceptionMapper;
+import io.quarkus.resteasy.runtime.CompositeExceptionMapper;
 import io.quarkus.resteasy.runtime.ExceptionMapperRecorder;
 import io.quarkus.resteasy.runtime.ForbiddenExceptionMapper;
 import io.quarkus.resteasy.runtime.JaxRsSecurityConfig;
@@ -73,6 +76,9 @@ public class ResteasyBuiltinsProcessor {
         providers.produce(new ResteasyJaxrsProviderBuildItem(UnauthorizedExceptionMapper.class.getName()));
         providers.produce(new ResteasyJaxrsProviderBuildItem(ForbiddenExceptionMapper.class.getName()));
         providers.produce(new ResteasyJaxrsProviderBuildItem(AuthenticationFailedExceptionMapper.class.getName()));
+        providers.produce(new ResteasyJaxrsProviderBuildItem(AuthenticationRedirectExceptionMapper.class.getName()));
+        providers.produce(new ResteasyJaxrsProviderBuildItem(AuthenticationCompletionExceptionMapper.class.getName()));
+        providers.produce(new ResteasyJaxrsProviderBuildItem(CompositeExceptionMapper.class.getName()));
         if (capabilities.isPresent(Capability.SECURITY)) {
             providers.produce(new ResteasyJaxrsProviderBuildItem(SecurityContextFilter.class.getName()));
             additionalBeanBuildItem.produce(AdditionalBeanBuildItem.unremovableOf(SecurityContextFilter.class));

--- a/extensions/resteasy/runtime/src/main/java/io/quarkus/resteasy/runtime/AuthenticationCompletionExceptionMapper.java
+++ b/extensions/resteasy/runtime/src/main/java/io/quarkus/resteasy/runtime/AuthenticationCompletionExceptionMapper.java
@@ -1,0 +1,17 @@
+package io.quarkus.resteasy.runtime;
+
+import javax.ws.rs.core.Response;
+import javax.ws.rs.ext.ExceptionMapper;
+import javax.ws.rs.ext.Provider;
+
+import io.quarkus.security.AuthenticationCompletionException;
+
+@Provider
+public class AuthenticationCompletionExceptionMapper implements ExceptionMapper<AuthenticationCompletionException> {
+
+    @Override
+    public Response toResponse(AuthenticationCompletionException ex) {
+        return Response.status(401).build();
+    }
+
+}

--- a/extensions/resteasy/runtime/src/main/java/io/quarkus/resteasy/runtime/AuthenticationRedirectExceptionMapper.java
+++ b/extensions/resteasy/runtime/src/main/java/io/quarkus/resteasy/runtime/AuthenticationRedirectExceptionMapper.java
@@ -1,0 +1,22 @@
+package io.quarkus.resteasy.runtime;
+
+import javax.ws.rs.core.HttpHeaders;
+import javax.ws.rs.core.Response;
+import javax.ws.rs.ext.ExceptionMapper;
+import javax.ws.rs.ext.Provider;
+
+import io.quarkus.security.AuthenticationRedirectException;
+
+@Provider
+public class AuthenticationRedirectExceptionMapper implements ExceptionMapper<AuthenticationRedirectException> {
+
+    @Override
+    public Response toResponse(AuthenticationRedirectException ex) {
+        return Response.status(ex.getCode())
+                .header(HttpHeaders.LOCATION, ex.getRedirectUri())
+                .header(HttpHeaders.CACHE_CONTROL, "no-store")
+                .header("Pragma", "no-cache")
+                .build();
+    }
+
+}

--- a/extensions/resteasy/runtime/src/main/java/io/quarkus/resteasy/runtime/CompositeExceptionMapper.java
+++ b/extensions/resteasy/runtime/src/main/java/io/quarkus/resteasy/runtime/CompositeExceptionMapper.java
@@ -1,0 +1,31 @@
+package io.quarkus.resteasy.runtime;
+
+import javax.ws.rs.InternalServerErrorException;
+import javax.ws.rs.core.Context;
+import javax.ws.rs.core.Response;
+import javax.ws.rs.ext.ExceptionMapper;
+import javax.ws.rs.ext.Provider;
+import javax.ws.rs.ext.Providers;
+
+import io.smallrye.mutiny.CompositeException;
+
+@Provider
+public class CompositeExceptionMapper implements ExceptionMapper<CompositeException> {
+
+    @Context
+    Providers p;
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public Response toResponse(CompositeException ex) {
+        Throwable t = ex.getCauses().stream().filter(s -> s != null).findFirst()
+                .orElseThrow(() -> new InternalServerErrorException());
+
+        ExceptionMapper<Throwable> mapper = (ExceptionMapper<Throwable>) p.getExceptionMapper(t.getClass());
+        if (mapper == null) {
+            throw new InternalServerErrorException();
+        }
+        return mapper.toResponse(t);
+    }
+
+}

--- a/integration-tests/oidc-code-flow/src/main/java/io/quarkus/it/keycloak/TenantLogout.java
+++ b/integration-tests/oidc-code-flow/src/main/java/io/quarkus/it/keycloak/TenantLogout.java
@@ -22,6 +22,16 @@ public class TenantLogout {
         return "Tenant Logout";
     }
 
+    // It is needed for the proactive-auth=false to work: /tenant-logout/logout should match a user initiated logout request 
+    // which must be handled by `CodeAuthenticationMechanism`.
+    // Adding `@Authenticated` gives control to `CodeAuthenticationMechanism` instead of RestEasy.
+    @GET
+    @Authenticated
+    @Path("logout")
+    public String getTenantLogoutPath() {
+        throw new InternalServerErrorException();
+    }
+
     @GET
     @Path("post-logout")
     public String postLogout(@QueryParam("state") String postLogoutState) {

--- a/integration-tests/oidc-code-flow/src/main/resources/application.properties
+++ b/integration-tests/oidc-code-flow/src/main/resources/application.properties
@@ -67,7 +67,7 @@ quarkus.oidc.tenant-3.auth-server-url=${keycloak.url}/realms/quarkus
 quarkus.oidc.tenant-3.client-id=quarkus-app
 quarkus.oidc.tenant-3.credentials.secret=secret
 quarkus.oidc.tenant-3.token.issuer=${keycloak.url}/realms/quarkus
-quarkus.oidc.tenant-3.authentication.redirect-path=/some/other/path
+quarkus.oidc.tenant-3.authentication.redirect-path=/web-app3
 quarkus.oidc.tenant-3.authentication.restore-path-after-redirect=false
 quarkus.oidc.tenant-3.application-type=web-app
 
@@ -114,6 +114,9 @@ quarkus.oidc.tenant-javascript.application-type=web-app
 quarkus.http.auth.permission.roles1.paths=/index.html
 quarkus.http.auth.permission.roles1.policy=authenticated
 
+quarkus.http.auth.permission.callback.paths=/web-app3
+quarkus.http.auth.permission.callback.policy=authenticated
+
 quarkus.http.auth.permission.logout.paths=/tenant-logout
 quarkus.http.auth.permission.logout.policy=authenticated
 
@@ -130,7 +133,7 @@ quarkus.http.auth.permission.post-logout.paths=/tenant-logout/post-logout
 quarkus.http.auth.permission.post-logout.policy=permit
 
 quarkus.http.cors=true
-
+quarkus.http.auth.proactive=false
 quarkus.log.category."io.quarkus.oidc.runtime.CodeAuthenticationMechanism".level=TRACE
 quarkus.log.category."io.vertx.ext.auth.oauth2.impl.OAuth2UserImpl".level=TRACE
 #quarkus.log.category."org.apache.http".level=TRACE

